Update “Thema maken” pagina met meer uitleg (#547)
nl-design-system/documentatie: docs/handboek/developer/03-thema-maken.mdx

De kleuren zetten we uiteindelijk in een JSON-bestand (in het [Design Tokens Format](https://tr.designtokens.org/format/)):

```json title="colors.tokens.json"
{
  "organisatie": {
    "color": {
      "red": {
        "30": { "value": "#911808" },
        "50": { "value": "#f2280d" },
        "70": { "value": "#f77e6e" }
      }
    }
  }
}
```

Dat bestand kan bijvoorbeeld zo in de filestructuur staan:

```bash
├─ src
  ├─ brand
    ├─ organisatie
      ├─ colors.tokens.json
```

Binnen NL Design System gebruiken we ook design tokens voor andere designkeuzes, waaronder:

- typografie (lettertypes, groottes en weights)
- spacing
- focus styles

### Componenten en variabelen

Als de algemene tokens zijn gemaakt, kan er ook per component een bestand met tokens worden toegevoegd, waarin ook naar de andere beschikbare tokens kan worden verwezen, zelfs in andere organisaties. Hier wordt bijvoorbeeld in de tokens voor het `link` component van `organisatie` verwezen naar een kleur rood van `andereOrganisatie`, met de verwijzing in curly brackets:

```json title="link.tokens.json"

> "organisatie": {

Reviewer: Is this an option? 

```json
{
  "organisatie": {
    "link": {
      "color": { "value": "{andereOrganisatie.color.red.45}" }
    }
  }
}
```

Author: Doh! Yes. <3

# NL Design System thema maken

Zoals beschreven in [Ontwikkel een thema voor jouw organisatie](/meedoen/als-developer/aan-de-slag#ontwikkel-een-thema-voor-jouw-organisatie): thema's zijn technisch JSON-bestanden waarin design tokens zijn gedefinieerd. Met tooling maken we ze beschikbaar als custom properties in CSS.

## Design tokens vastleggen

> Design tokens kunnen allerlei ontwerpkeuzes bevatten: kleuren, lettergroottes, afstanden en meer. Om een theme te maken vanuit een bestaande website en/of huisstijl maken we vaak een inventarisatie, met als einddoel één of meer JSON-bestanden.

Reviewer: theme -> thema?

Author: ```suggestion
Design tokens kunnen allerlei ontwerpkeuzes bevatten: kleuren, lettergroottes, afstanden en meer. Om een thema te maken vanuit een bestaande website en/of huisstijl maken we vaak een inventarisatie, met als einddoel één of meer JSON-bestanden.
```

Het kan ook zijn dat er duidelijk meerdere varianten zijn van één kleur, bijvoorbeeld drie tinten rood. Je kunt tokens noemen hoe je wilt, het is vaak handig om er iets van een systeem in aan te brengen. Soms kan dat uit een huisstijlgids worden overgenomen. Als er geen bestaand systeem is in de namen, kun je ook de kleurwaarde naar [HSL](https://hslpicker.com) converteren en dan de lightness gebruiken (de “L” uit “HSL”), zodat je bv met `rood-30`, `rood-50` en `rood-70`. Rond gerust af op hele getallen.

De kleuren zetten we uiteindelijk in een JSON-bestand (in het [Design Tokens Format](https://tr.designtokens.org/format/)):

```json title="colors.tokens.json"
{
  "organisatie": {
    "color": {
      "red": {
        "30": { "value": "#911808" },
        "50": { "value": "#f2280d" },
        "70": { "value": "#f77e6e" }
      }
    }
  }
}
```

Dat bestand kan bijvoorbeeld zo in de filestructuur staan:

```txt
├─ src
  ├─ brand
    ├─ organisatie
      ├─ colors.tokens.json
```

Binnen NL Design System gebruiken we ook design tokens voor andere designkeuzes, waaronder:

> - typografie (lettertypes, groottes en weights)

Reviewer: weights -> gewichten, of juist alles Engels?

Author: ```suggestion
- typografie (lettertypes, groottes en gewichten)
```

- spacing
- focus styles

### Componenten en variabelen

> Als de algemene tokens zijn gemaakt, kan er ook per component een bestand met tokens worden toegevoegd, waarin ook naar de andere beschikbare tokens kan worden verwezen, zelfs in andere organisaties. Hier wordt bijvoorbeeld in de tokens voor het `link` component van `organisatie` verwezen naar een kleur rood van `andereOrganisatie`, met de verwijzing in curly brackets:

Reviewer: Hier -> in het volgende voorbeeld?

Author: ```suggestion
Als de algemene tokens zijn gemaakt, kan er ook per component een bestand met tokens worden toegevoegd, waarin ook naar de andere beschikbare tokens kan worden verwezen, zelfs in andere organisaties. In het volgende voorbeeld wordt bijvoorbeeld in de tokens voor het `link` component van `organisatie` verwezen naar een kleur rood van `andereOrganisatie`, met de verwijzing in curly brackets:
```

```json title="link.tokens.json"
{
  "organisatie": {
    "link": {
      "color": { "value": "{andereOrganisatie.color.red.45}" }
    }
  }
}
```

### Meer over design tokens vastleggen

<VideoPlayer videoId="jSvjnU7oY5Y" />

## Design tokens inzetten

Zoals genoemd kunnen deze JSON-bestanden in allerlei toepassingen worden gebruikt. Bij NL Design System zorgt [Style Dictionary](https://amzn.github.io/style-dictionary) dat de tokens bruikbaar worden in CSS, door een `design-tokens.css` te genereren. Die output kan er bijvoorbeeld zo uit zien:

```css title="dist/design-tokens.css"
/**
  * Do not edit directly
  * Generated at […]
  */
.organisatie-theme {
  --organisatie-color-red-30: #911808;
  --organisatie-color-red-50: #f2280d;
  --organisatie-color-red-70: #f77e6e;
}
```

> In dit bestand worden alle tokens die in het theme van deze organisatie bestaan vanuit verschillende plekken bij elkaar gezet, als custom properties. Nu kunnen ze in de CSS van deze organisatie worden gebruikt, op pagina's waar de `organisatie-theme` classname op de body staat. Er is ook een output setting waar de variabelen op `:root` worden geplaatst, voor als de HTML niet makkelijk aanpasbaar is.

Reviewer: Niet alleen op de body... maar ik weet niet hoe dat beter te verwoorden. Misschien de body als voorbeeld noemen? Of aangeven dat dat de gebruikelijke plek is?

Author: zoiets?

```suggestion
In dit bestand worden alle tokens die in het theme van deze organisatie bestaan vanuit verschillende plekken bij elkaar gezet, als custom properties. Nu kunnen ze in de CSS van deze organisatie worden gebruikt, op pagina's waar de `organisatie-theme` classname wordt gebruikt, meestal op de `body`. Er is ook een output setting waar de variabelen op `:root` worden geplaatst, voor als de HTML niet makkelijk aanpasbaar is.
```

Dat hoeft niet altijd één op één. Soms valt dan op dat sommige kleuren wel erg dicht bij elkaar liggen, dan kan bijvoorbeeld de meest voorkomende gekozen worden en de rest afgerond worden, de afwijking is dan vast eerder een foutje dan een keuze.

Het kan ook zijn dat er duidelijk meerdere varianten zijn van één kleur, bijvoorbeeld drie tinten rood. Je kunt tokens noemen hoe je wilt, het is vaak handig om er iets van een systeem in aan te brengen. Soms kan dat uit een huisstijlgids worden overgenomen. Als er geen bestaand systeem is in de namen, kun je ook de kleurwaarde naar [HSL](https://hslpicker.com) converteren en dan de lightness gebruiken (de “L” uit “HSL”), zodat je bv met `rood-30`, `rood-50` en `rood-70`. Rond gerust af op hele getallen.

De kleuren zetten we uiteindelijk in een JSON-bestand (in het [Design Tokens Format](https://tr.designtokens.org/format/)):

```js
// colors.tokens.json
{
  "organisatie": {
    "color": {
      "red": {
        "30": { "value": "#911808" },
        "50": { "value": "#f2280d" },
        "70": { "value": "#f77e6e" }
      }
    }
  }
}
```

Dat bestand kan bijvoorbeeld zo in de filestructuur staan:

> ```bash

Reviewer: https://github.com/github-linguist/linguist/blob/master/lib/linguist/languages.yml#L7092

`plain text` apparently…
